Mathlib/RingTheory/AdicCompletion/RingHom.lean

```diff
@@ -0,0 +1,282 @@
+    (ofAlgEquiv I).symm (of I S x) = x := by
+  simp [ofAlgEquiv]
+
+theorem mk_smul_top_ofAlgEquiv_symm (n : ℕ) (x : AdicCompletion I S) :
+    Ideal.Quotient.mk (I ^ n • ⊤) ((ofAlgEquiv I).symm x) = eval I S n x := by
+  nth_rw 2 [← of_ofAlgEquiv_symm I x]
+  simp [-of_ofAlgEquiv_symm, eval,-smul_eq_mul,-Algebra.id.smul_eq_mul]
+
+@[simp]
+theorem mk_ofAlgEquiv_symm (n : ℕ) (x : AdicCompletion I S) :
+    Ideal.Quotient.mk (I ^ n) ((ofAlgEquiv I).symm x) = evalₐ I n x := by
+  simp only [evalₐ, AlgHom.coe_comp, Function.comp_apply, AlgHom.ofLinearMap_apply]
+  rw [← mk_smul_top_ofAlgEquiv_symm I n x]
+  simp
+
+end AdicCompletion
+
+namespace IsAdicComplete
+
+open AdicCompletion
+
+section
+
+variable [IsAdicComplete I S] (f : (n : ℕ) → R →+* S ⧸ I ^ n)
+    (hf : ∀ {m n : ℕ} (hle : m ≤ n), (factorPow I hle).comp (f n) = f m)
+
+/--
+Universal property of `IsAdicComplete` for rings.
+The lift ring map `lift I f hf : R →+* S` of a sequence of compatible
+ring maps `f n : R →+* S ⧸ (I ^ n)`.
+-/
+noncomputable def liftRingHom :
+    R →+* S :=
+  ((ofAlgEquiv I).symm : _ →+* _).comp (AdicCompletion.liftRingHom I f hf)
+
+@[simp]
+theorem of_liftRingHom (x : R) :
+    of I S (liftRingHom I f hf x) = (AdicCompletion.liftRingHom I f hf x) := by
+  simp [liftRingHom]
+
+@[simp]
+theorem ofAlgEquiv_comp_liftRingHom :
+    (ofAlgEquiv I : S →+* AdicCompletion I S).comp (liftRingHom I f hf) =
+      AdicCompletion.liftRingHom I f hf := by
+  ext; simp
+
+/--
+The composition of lift linear map `lift I f hf : R →+* S` with the canonical
+projection `S →+* S ⧸ (I ^ n)` is `f n` .
+-/
+@[simp]
+theorem mk_liftRingHom (n : ℕ) (x : R) :
+    Ideal.Quotient.mk (I ^ n) (liftRingHom I f hf x) = f n x:= by
+  simp only [liftRingHom, RingHom.coe_comp, RingHom.coe_coe, Function.comp_apply]
+  rw [← evalₐ_of_apply I n]
+  simp
+
+@[simp]
+theorem mk_comp_liftRingHom (n : ℕ) :
+    (Ideal.Quotient.mk (I ^ n)).comp (liftRingHom I f hf) = f n := by
+  ext; simp
+
+/--
+Uniqueness of the lift.
+Given a compatible family of linear maps `f n : R →ₗ[R] S ⧸ (I ^ n)`.
+If `F : R →+* S` makes the following diagram commutes
+```
+  R
+  | \
+ F|  \ f n
+  |   \
+  v    v
+  S --> S ⧸ (I ^ n)
+```
+Then it is the map `IsAdicComplete.lift`.
+-/
+theorem eq_liftRingHom (F : R →+* S)
+    (hF : ∀ n, (Ideal.Quotient.mk (I ^ n)).comp F = f n) :
+    F = liftRingHom I f hf := by
+  ext x
+  rw [IsHausdorff.eq_iff_smodEq (I := I)]
+  intro n
+  simp only [smul_eq_mul, mul_top]
+  simp [SModEq, liftRingHom, ← hF n]
+
+end
+
+namespace StrictMono
+
+variable {a : ℕ → ℕ} (ha : StrictMono a)
+    (f : (n : ℕ) → R →+* S ⧸ I ^ a n)
+
+variable {I M}
+/--
+Instead of providing all `R →+* S ⧸ I ^ n`, one can just provide
+`R →+* S ⧸ I ^ a n` for a strictly increasing sequence `a n` to recover all
+`R →+* S ⧸ I ^ n`. `RingHom` variant of `IsAdicComplete.StrictMono.extend`.
+-/
+def extendRingHom (n : ℕ) :
+    R →+* S ⧸ I ^ n :=
+  (factorPow I (ha.id_le n)).comp (f n)
```

Comment:
I would prefer to not have this def and write `(factorPow I (ha.id_le n)).comp (f n)` inline.